### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by fzzylogic** [Jermolene/TiddlyWiki5#6018]
**Describe the bug**
Using Codemirror plugin, shortcuts that overlap with browser shortcuts, activate browser features, even when focus was in the tiddler edit area.

**To Reproduce**
0. Download a fresh version of Tiddlywiki, taking not of the published version number.
1. Install Codemirror plugin.
2. In Firefox / Chrome / MS Edge (tested), go to a new tiddler.
3. With focus in tiddler text area, use a shortcut not present in the browser (e.g. ALT-G). The codemirror line selection field should display.
4. Use a shortcut present in the browser (e.g. ALT-F). The browser feature is activated instead of the codemirror function.

**Expected behavior**
When the tiddler text field has focus, Codemirror shortcuts should take precedence.

**Screenshots**
N/A

**TiddlyWiki Configuration (please complete the following information):**
 - Version [e.g. v5.1.23] [I see what you did there, clever :-) ]
 - Saving mechanism (the underrated save button)
 - Plugins installed [e.g. Codemirror]
 
**Desktop (please complete the following information):**
 - OS: [e.g. Windows 10]
 - Browser [e.g. Chrome / Firefox / MS Edge]
 - Version [e.g. latest]

**Smartphone (please complete the following information):**
N/A

**Additional context**
Tiddlywiki is fantastic btw.

**Comment by Jermolene** [comment]
Thanks @fzzylogic

@pmario @BurningTreeC do you know what's going on here?

**Comment by BurningTreeC** [comment]
Hi @fzzylogic , @Jermolene 

If I try a keyboard shortcut like <kbd>ctrl-T</kbd> which should apply strikethrough formatting, in both the codemirror and the framed editor it does nothing and opens a browser tab instead. So the problem might be deeper. Tested this in chrome/chromium and firefox

For the codemirror-specific shortcuts I'll have a look

**Comment by pmario** [comment]
`CTRL-T` I personally always thought, that we can't prevent, that shortcuts that are catched by the browser can be overwritten. I do reconfigure them in TW if I need them.

**Comment by pmario** [comment]
see: https://stackoverflow.com/questions/38838302/any-way-to-override-ctrln-to-open-a-new-window-in-chrome

**Comment by BurningTreeC** [comment]
Yes @pmario, I also think that the problem here is that pages cannot override the shortcuts catched by the browser

**Comment by BurningTreeC** [comment]
@pmario @fzzylogic @Jermolene ,

the solution for the codemirror editor would be the following:

- identify the name of the command that's intercepted by the browser, looking [on this page](https://codemirror.net/doc/manual.html#commands)
- open the tiddler `$:/config/codemirror/extraKeysTW` for editing
- add a different shortcut, one that is not used by the browser, for the command

like the following example with the <kbd>Alt-F</kbd> command which (I think) should be `goWordRight`:

```json
{
	"Ctrl-Esc": "singleSelection",
	"Esc": "",
	"Ctrl-S": "",
	"Ctrl-U": "",
	"Ctrl-T": "",
	"Alt-T": "transposeChars",
	"Alt-U": "undoSelection",
	"Shift-Alt-U": "redoSelection",
	"Cmd-U": "",
	"Tab": "indentAuto()",
	"Enter": "newLineAndIndent()",
	"Shift-Alt-F": "goWordRight()"
}
```

**Comment by fzzylogic** [comment]
Thank you so much everyone for looking at this and for the workaround for replacing shortcuts. I did wonder whether browsers will always enforce their shortcuts, but without looking into it too deeply, noticed that it is possible to use browser shortcuts, at least while the focus is on the right element. For example, see [ACE editor](https://ace.c9.io/), [Monaco Editor](https://microsoft.github.io/monaco-editor/playground.html) etc.

**Comment by fzzylogic** [comment]
tl;dr: Putting shortcuts in extraKeysTW works, and is able to override browser default shortcuts, but not tiddlywiki defaults (probably by design).

Noticed a few things while looking at this, just putting the info here. 

- ALT-F is successfully intercepted if set in $:/config/codemirror/extraKeysTW, if *used without the closing brackets*:

    {"Alt-F": "goWordRight"}

- This doesn't work within extraKeysTW tiddler unless the type is changed from 'json' to 'application/json'.
- Alt-F is mentioned in plugin "CodeMirror Search and Replace" as being for "Persistent search", don't see that option in the codemirror commands, so not sure how to activate that.
- Shift-Ctrl-F is mentioned in Search and Replace plugin as being for "Replace", but it doesn't work, apparently because it would override a tiddlywiki shortcut (the one that sends the cursor to the main search field of the sidebar).
- Mapping Ctrl-H to replace in extraKeysTW works nicely :-)
- Previously had the Sublime plugin installed which mentions Ctrl-H as the key-mapping for replace (implemented in $:/plugins/tiddlywiki/codemirror/keymap/sublime.js, not extraKeysTW), but that doesn't work.

**Comment by pmario** [comment]
@Jermolene @BurningTreeC .. may be CM uses a different mechanism than TW to "intercept" keyboard shortcuts. .. We may have a closer look, how `extraKeysTW` does the trick.

**Comment by fzzylogic** [comment]
In line with @BurningTreeC's comment above, have also learned that it is possible to override many browser shortcuts, but not all (and of course not OS shortcuts). For example, shortcuts for actions on tabs or windows can't be overridden. Here are some, couldn't find a comprehensive list:

    - Ctrl+N
    - Ctrl+Shift+N
    - Ctrl+T
    - Ctrl-Tab
    - Ctrl-Shift-Tab
    - Ctrl+Shift+T
    - Ctrl+Shift+P
    - Ctrl-Shift-Q (closes every tab and window without warning in Chrome / Firefox)
    - Ctrl+W
    - Ctrl+PgUp/PgDn

Made the below to make testing more convenient, it tries to intercept everything.

```
<html><head></head><body contenteditable="true">
Open dev console, put cursor here and press shortcuts to see what can be intercepted.
<script>
function key_intercept(e) {
    var keys = []
    e.ctrlKey ? keys.push('Ctrl') : null
    e.altKey ? keys.push('Alt') : null
    e.shiftKey ? keys.push('Shift') : null
    if (e.key !== 'Control' && e.key !== 'Alt' && e.key !== 'Shift'){
      keys.push(e.key)
    }
    e.preventDefault()
    var d = document.createElement("div")
    d.textContent = "Intercepted: " + keys.join('+')
    document.body.insertBefore(d, document.body.firstChild)
}
document.addEventListener('keydown', key_intercept, false);
</script></body></html>
```

**Comment by pmario** [comment]
@fzzylogic -- If the problem is resolved for you can you close the issue?
